Add checksum information on the release page

twang2218 opened the pull request:
Fixes #29515



**- What I did**

Add both checksum and links of `sha256sum` and `md5sum` to the release
page.

**- How I did it**

* Extract 2 functions, `map_goos_to_uname()` and `map_goarch_to_uname()`, to reuse.
* Add 2 functions, `get_release_entry()` and `get_release_list()` for generating the information of the release binary link, checksum links, and checksum values.
* Get the release information before the files were removed.

**- How to verify it**

I tested it with following command:

```bash
$ make
$ docker run -e AWS_S3_BUCKET=docker.lab99.org \
            -e AWS_ACCESS_KEY_ID \
            -e AWS_SECRET_ACCESS_KEY \
            -e AWS_DEFAULT_REGION \
            -it --privileged \
            docker-dev:add-checksum-info \
            ./hack/release.sh --release-regardless-of-test-failure
```

and generate following markdown:

```markdown
Release complete; see http://docker.lab99.org.s3-website-ap-southeast-2.amazonaws.com
Use the following text to announce the release:

We have just pushed 1.14.0-dev to http://docker.lab99.org.s3-website-ap-southeast-2.amazonaws.com. You can download it with the following:

Linux 64bit tgz: http://docker.lab99.org.s3-website-ap-southeast-2.amazonaws.com/builds/Linux/x86_64/docker-1.14.0-dev.tgz
* sha256sum: [`882f900822cfe45e33d2c3788267587bc0b8c21513c680e8150f53a896c2cdea`](http://docker.lab99.org.s3-website-ap-southeast-2.amazonaws.com/builds/Linux/x86_64/docker-1.14.0-dev.tgz.sha256)
* md5sum: [`be2785441d24e6a8a4f9273dc445d2f9`](http://docker.lab99.org.s3-website-ap-southeast-2.amazonaws.com/builds/Linux/x86_64/docker-1.14.0-dev.tgz.md5)

Linux 32bits arm tgz: http://docker.lab99.org.s3-website-ap-southeast-2.amazonaws.com/builds/Linux/armel/docker-1.14.0-dev.tgz
* sha256sum: [`7c01a295d9c3be300fdca0df13918ef063d19d364c8c6b0623bb0512516c1938`](http://docker.lab99.org.s3-website-ap-southeast-2.amazonaws.com/builds/Linux/armel/docker-1.14.0-dev.tgz.sha256)
* md5sum: [`66fee6a0b1bc8f2d24081fc1a5802b7d`](http://docker.lab99.org.s3-website-ap-southeast-2.amazonaws.com/builds/Linux/armel/docker-1.14.0-dev.tgz.md5)

Darwin/OSX 64bit client tgz: http://docker.lab99.org.s3-website-ap-southeast-2.amazonaws.com/builds/Darwin/x86_64/docker-1.14.0-dev.tgz
* sha256sum: [`969fdafdcec562739149343485555e70c7a90b334888a26c5960b940c2eccc5a`](http://docker.lab99.org.s3-website-ap-southeast-2.amazonaws.com/builds/Darwin/x86_64/docker-1.14.0-dev.tgz.sha256)
* md5sum: [`d6bc77d49ae2326600abc551b49f8e56`](http://docker.lab99.org.s3-website-ap-southeast-2.amazonaws.com/builds/Darwin/x86_64/docker-1.14.0-dev.tgz.md5)

Windows 64bit zip: http://docker.lab99.org.s3-website-ap-southeast-2.amazonaws.com/builds/Windows/x86_64/docker-1.14.0-dev.zip
* sha256sum: [`695f704640fac71ec43fbe56444c353acb4261988394850f8cbb12f62c5dee39`](http://docker.lab99.org.s3-website-ap-southeast-2.amazonaws.com/builds/Windows/x86_64/docker-1.14.0-dev.zip.sha256)
* md5sum: [`c6df82ed3c22c7d94c38ab195d0bddd4`](http://docker.lab99.org.s3-website-ap-southeast-2.amazonaws.com/builds/Windows/x86_64/docker-1.14.0-dev.zip.md5)

Windows 32bit client zip: http://docker.lab99.org.s3-website-ap-southeast-2.amazonaws.com/builds/Windows/i386/docker-1.14.0-dev.zip
* sha256sum: [`86de7d0b6508352b35ed68590dba32c1dccad093d2bd82cd874704bdc331fee7`](http://docker.lab99.org.s3-website-ap-southeast-2.amazonaws.com/builds/Windows/i386/docker-1.14.0-dev.zip.sha256)
* md5sum: [`aa3ff350c9df43fc8a5e660eb3118466`](http://docker.lab99.org.s3-website-ap-southeast-2.amazonaws.com/builds/Windows/i386/docker-1.14.0-dev.zip.md5)
```

The rendered result is:

---
Release complete; see http://docker.lab99.org.s3-website-ap-southeast-2.amazonaws.com
Use the following text to announce the release:

We have just pushed 1.14.0-dev to http://docker.lab99.org.s3-website-ap-southeast-2.amazonaws.com. You can download it with the following:

Linux 64bit tgz: http://docker.lab99.org.s3-website-ap-southeast-2.amazonaws.com/builds/Linux/x86_64/docker-1.14.0-dev.tgz
* sha256sum: [`882f900822cfe45e33d2c3788267587bc0b8c21513c680e8150f53a896c2cdea`](http://docker.lab99.org.s3-website-ap-southeast-2.amazonaws.com/builds/Linux/x86_64/docker-1.14.0-dev.tgz.sha256)
* md5sum: [`be2785441d24e6a8a4f9273dc445d2f9`](http://docker.lab99.org.s3-website-ap-southeast-2.amazonaws.com/builds/Linux/x86_64/docker-1.14.0-dev.tgz.md5)

Linux 32bits arm tgz: http://docker.lab99.org.s3-website-ap-southeast-2.amazonaws.com/builds/Linux/armel/docker-1.14.0-dev.tgz
* sha256sum: [`7c01a295d9c3be300fdca0df13918ef063d19d364c8c6b0623bb0512516c1938`](http://docker.lab99.org.s3-website-ap-southeast-2.amazonaws.com/builds/Linux/armel/docker-1.14.0-dev.tgz.sha256)
* md5sum: [`66fee6a0b1bc8f2d24081fc1a5802b7d`](http://docker.lab99.org.s3-website-ap-southeast-2.amazonaws.com/builds/Linux/armel/docker-1.14.0-dev.tgz.md5)

Darwin/OSX 64bit client tgz: http://docker.lab99.org.s3-website-ap-southeast-2.amazonaws.com/builds/Darwin/x86_64/docker-1.14.0-dev.tgz
* sha256sum: [`969fdafdcec562739149343485555e70c7a90b334888a26c5960b940c2eccc5a`](http://docker.lab99.org.s3-website-ap-southeast-2.amazonaws.com/builds/Darwin/x86_64/docker-1.14.0-dev.tgz.sha256)
* md5sum: [`d6bc77d49ae2326600abc551b49f8e56`](http://docker.lab99.org.s3-website-ap-southeast-2.amazonaws.com/builds/Darwin/x86_64/docker-1.14.0-dev.tgz.md5)

Windows 64bit zip: http://docker.lab99.org.s3-website-ap-southeast-2.amazonaws.com/builds/Windows/x86_64/docker-1.14.0-dev.zip
* sha256sum: [`695f704640fac71ec43fbe56444c353acb4261988394850f8cbb12f62c5dee39`](http://docker.lab99.org.s3-website-ap-southeast-2.amazonaws.com/builds/Windows/x86_64/docker-1.14.0-dev.zip.sha256)
* md5sum: [`c6df82ed3c22c7d94c38ab195d0bddd4`](http://docker.lab99.org.s3-website-ap-southeast-2.amazonaws.com/builds/Windows/x86_64/docker-1.14.0-dev.zip.md5)

Windows 32bit client zip: http://docker.lab99.org.s3-website-ap-southeast-2.amazonaws.com/builds/Windows/i386/docker-1.14.0-dev.zip
* sha256sum: [`86de7d0b6508352b35ed68590dba32c1dccad093d2bd82cd874704bdc331fee7`](http://docker.lab99.org.s3-website-ap-southeast-2.amazonaws.com/builds/Windows/i386/docker-1.14.0-dev.zip.sha256)
* md5sum: [`aa3ff350c9df43fc8a5e660eb3118466`](http://docker.lab99.org.s3-website-ap-southeast-2.amazonaws.com/builds/Windows/i386/docker-1.14.0-dev.zip.md5)
---

**- A picture of a cute animal (not mandatory but encouraged)**

A cute wombat.

![cute-wombat](https://cloud.githubusercontent.com/assets/6299096/21384342/4e5932d0-c7be-11e6-936a-3abda2d54359.jpg)


Signed-off-by: Tao Wang <twang2218@gmail.com>

twang2218 committed:
Add checksum information on the release page

Fixes #29515

Add `sha256sum` info to the release page.

Signed-off-by: Tao Wang <twang2218@gmail.com>

thaJeztah commented:
ping @tianon @andrewhsu PTAL

twang2218 commented:
Any updates?

thaJeztah commented:
Let me ask @tianon again to have a look :heart:

tianon commented on hack/release.sh:
```diff
@@ -114,6 +114,92 @@ s3_url() {
+	# see https://en.wikipedia.org/wiki/Uname
+	# ie, GOOS=linux -> "uname -s"=Linux
+	GOOS=$1
+	case "$GOOS" in
```
Any particular reason to do this in two lines rather than just one? ie:

```bash
map_goos_to_uname() {
	# we need to map our GOOS to uname values
	# see https://en.wikipedia.org/wiki/Uname
	# ie, GOOS=linux -> "uname -s"=Linux
	case "$1" in
...
```

twang2218 replied:
I think it's has clearer meaning of using a named local variable for the argument, but it's fine to use `$1` as it's the only argument of the function. I will change to single line style.

tianon commented on hack/release.sh:
```diff
@@ -114,6 +114,92 @@ s3_url() {
+	# see https://en.wikipedia.org/wiki/Uname
+	# ie, GOARCH=amd64 -> "uname -m"=x86_64
+	GOARCH=$1
+	case "$GOARCH" in
```
Same thought here -- these might as well be one line and not create a new variable (which might modify a variable unintentionally too, since it's not `local`).

tianon commented on hack/release.sh:
```diff
@@ -114,6 +114,92 @@ s3_url() {
+			echo "Windows"
+			;;
+		*)
+			echo >&2 "error: can't convert $GOOS to an appropriate value for 'uname -s'"
```
Error would need to be updated to use `$1` as well. :+1:

tianon commented on hack/release.sh:
```diff
@@ -114,6 +114,92 @@ s3_url() {
+	GOARCH=$2
+	s3Os=`map_goos_to_uname $GOOS`
+	s3Arch=`map_goarch_to_uname $GOARCH`
+	zipExt=".tgz"
```
All these variables ought to be `local`, and `$()` should be preferred over backticks, ie:
```bash
get_release_entry() {
	local GOOS="$1"
	local GOARCH="$2"
	local s3Os="$(map_goos_to_uname "$GOOS")"
	local s3Arch="$(map_goarch_to_uname "$GOARCH")"
	local zipExt='.tgz'
	...
```

twang2218 replied:
ok, I will fix this.

tianon reviewed (commented):
Overall, I feel this is a tad bit verbose -- IMO, the md5sum isn't really worth advertising anymore, especially since this list gets copied into quite a few places IIRC (the release notes on GitHub and several mailing lists, where Markdown isn't applicable, so the link/formatting will be lost anyhow), which is why the current version is single-spaced and doesn't mention ARM, for example.

I _do_ think there'd be value in having a docs or download page somewhere that has a lot of this more verbose information, but IMO the release notes isn't a good place for adding that sort of detail (https://golang.org/dl is a decent example of good balance -- release notes simply link there for the heavy details).

If we do go for this (ie, other maintainers feel more strongly about adding it than I do), then I think we should probably stick to the same short list of platforms (maybe even drop 32bit Windows for consistency) and only list the SHA256.

twang2218 commented:
I removed the md5sum result in the release notes, fixed local variable issues, and removed the markdown syntax in the notes. I didn't remove the md5sum file generation, just removed them from the release message.

now the output is:

```bash
Release complete; see http://docker.lab99.org.s3-website-ap-southeast-2.amazonaws.com
Use the following text to announce the release:

We have just pushed 1.14.0-dev to http://docker.lab99.org.s3-website-ap-southeast-2.amazonaws.com. You can download it with the following:

Linux 64bit tgz: http://docker.lab99.org.s3-website-ap-southeast-2.amazonaws.com/builds/Linux/x86_64/docker-1.14.0-dev.tgz
* sha256sum: 84a38abc5319fb5afdcfcb57c7be21e64a4e648473dace8c2b4dfd76b4b3a31a

Linux 32bits arm tgz: http://docker.lab99.org.s3-website-ap-southeast-2.amazonaws.com/builds/Linux/armel/docker-1.14.0-dev.tgz
* sha256sum: d07f1d17180472e37f68b39a1d9497a41cb973ba6fce20846086f709b4d51d4f

Darwin/OSX 64bit client tgz: http://docker.lab99.org.s3-website-ap-southeast-2.amazonaws.com/builds/Darwin/x86_64/docker-1.14.0-dev.tgz
* sha256sum: 9d22fa5b75daf2237912a171fa9a7a753646a3c885a5272a98d38be9acf90227

Windows 64bit zip: http://docker.lab99.org.s3-website-ap-southeast-2.amazonaws.com/builds/Windows/x86_64/docker-1.14.0-dev.zip
* sha256sum: 9d782573ba2449eb31e064ea5a51b5e104cd296a1c60d9346f8b53d0eaa636d0

Windows 32bit client zip: http://docker.lab99.org.s3-website-ap-southeast-2.amazonaws.com/builds/Windows/i386/docker-1.14.0-dev.zip
* sha256sum: 9449586775cb2cc72e9ecf02b70f081679ed8fd0926dfd6af1e6b32cee404984
```

And the markdown rendered result would be:

---
Release complete; see http://docker.lab99.org.s3-website-ap-southeast-2.amazonaws.com
Use the following text to announce the release:

We have just pushed 1.14.0-dev to http://docker.lab99.org.s3-website-ap-southeast-2.amazonaws.com. You can download it with the following:

Linux 64bit tgz: http://docker.lab99.org.s3-website-ap-southeast-2.amazonaws.com/builds/Linux/x86_64/docker-1.14.0-dev.tgz
* sha256sum: 84a38abc5319fb5afdcfcb57c7be21e64a4e648473dace8c2b4dfd76b4b3a31a

Linux 32bits arm tgz: http://docker.lab99.org.s3-website-ap-southeast-2.amazonaws.com/builds/Linux/armel/docker-1.14.0-dev.tgz
* sha256sum: d07f1d17180472e37f68b39a1d9497a41cb973ba6fce20846086f709b4d51d4f

Darwin/OSX 64bit client tgz: http://docker.lab99.org.s3-website-ap-southeast-2.amazonaws.com/builds/Darwin/x86_64/docker-1.14.0-dev.tgz
* sha256sum: 9d22fa5b75daf2237912a171fa9a7a753646a3c885a5272a98d38be9acf90227

Windows 64bit zip: http://docker.lab99.org.s3-website-ap-southeast-2.amazonaws.com/builds/Windows/x86_64/docker-1.14.0-dev.zip
* sha256sum: 9d782573ba2449eb31e064ea5a51b5e104cd296a1c60d9346f8b53d0eaa636d0

Windows 32bit client zip: http://docker.lab99.org.s3-website-ap-southeast-2.amazonaws.com/builds/Windows/i386/docker-1.14.0-dev.zip
* sha256sum: 9449586775cb2cc72e9ecf02b70f081679ed8fd0926dfd6af1e6b32cee404984
---


Is that link too long? should I put it in separate line as checksum does? Such as:

```bash
Windows 64bit zip: 
* Download: http://docker.lab99.org.s3-website-ap-southeast-2.amazonaws.com/builds/Windows/x86_64/docker-1.14.0-dev.zip
* sha256sum: 9d782573ba2449eb31e064ea5a51b5e104cd296a1c60d9346f8b53d0eaa636d0
```

Which would rendered as:

---
Windows 64bit zip: 
* Download: http://docker.lab99.org.s3-website-ap-southeast-2.amazonaws.com/builds/Windows/x86_64/docker-1.14.0-dev.zip
* sha256sum: 9d782573ba2449eb31e064ea5a51b5e104cd296a1c60d9346f8b53d0eaa636d0
---

thaJeztah commented:
ping @andrewhsu @mstanleyjones PTAL

mstanleyjones commented:
Instead of a bullet under each download link, why not in parentheses after each download link, or just (md5sum or sha256sum) that link to the checksum files, instead of displaying the checksums themselves? So it would look like:

Windows 64bit zip:

- Download: http://docker.lab99.org.s3-website-ap-southeast-2.amazonaws.com/builds/Windows/x86_64/docker-1.14.0-dev.zip ([sha256sum](link/to/sha256sum_file) or [md5sum](link/to/md5sum_file)))

thaJeztah commented:
I think there's value in printing the actual checksum, as these instructions are also sent per email; printing the checksum allows verifying the downloaded content (because both checksum and binaries are on the same host, both could theoretically be compromised)

I do agree that we can omit md5, and just use sha256

twang2218 commented:
Now I put the sha256sum result in the parentheses, and the sample result become:

```bash
Linux 64bit tgz: http://docker.lab99.org.s3-website-ap-southeast-2.amazonaws.com/builds/Linux/x86_64/docker-1.14.0-dev.tgz (sha256sum: 2a0c689b01d688d0cd5895ef79cc2c4a9fc8187e41f91a2576ba92d5da4531ca)

Linux 32bits arm tgz: http://docker.lab99.org.s3-website-ap-southeast-2.amazonaws.com/builds/Linux/armel/docker-1.14.0-dev.tgz (sha256sum: 38e332b3f0021631af4d434167f9fe370a80ac6bc86a91cd56ce0531c6dd062f)

Darwin/OSX 64bit client tgz: http://docker.lab99.org.s3-website-ap-southeast-2.amazonaws.com/builds/Darwin/x86_64/docker-1.14.0-dev.tgz (sha256sum: db9027abe8f2d087ca1aa6e8df2ba3e7921fd6326b040397aedcea9c5e9ca53e)

Windows 64bit zip: http://docker.lab99.org.s3-website-ap-southeast-2.amazonaws.com/builds/Windows/x86_64/docker-1.14.0-dev.zip (sha256sum: 96330f92c7f7960f20ebb0f59143739361727e9f1f6c4257387c0a712c097858)

Windows 32bit client zip: http://docker.lab99.org.s3-website-ap-southeast-2.amazonaws.com/builds/Windows/i386/docker-1.14.0-dev.zip (sha256sum: 45a1d18b1329e92064848af91af0a342a1c522eb6baf529ded1add587b896ced)
```

the markdown rendered result is:

---
Linux 64bit tgz: http://docker.lab99.org.s3-website-ap-southeast-2.amazonaws.com/builds/Linux/x86_64/docker-1.14.0-dev.tgz (sha256sum: 2a0c689b01d688d0cd5895ef79cc2c4a9fc8187e41f91a2576ba92d5da4531ca)

Linux 32bits arm tgz: http://docker.lab99.org.s3-website-ap-southeast-2.amazonaws.com/builds/Linux/armel/docker-1.14.0-dev.tgz (sha256sum: 38e332b3f0021631af4d434167f9fe370a80ac6bc86a91cd56ce0531c6dd062f)

Darwin/OSX 64bit client tgz: http://docker.lab99.org.s3-website-ap-southeast-2.amazonaws.com/builds/Darwin/x86_64/docker-1.14.0-dev.tgz (sha256sum: db9027abe8f2d087ca1aa6e8df2ba3e7921fd6326b040397aedcea9c5e9ca53e)

Windows 64bit zip: http://docker.lab99.org.s3-website-ap-southeast-2.amazonaws.com/builds/Windows/x86_64/docker-1.14.0-dev.zip (sha256sum: 96330f92c7f7960f20ebb0f59143739361727e9f1f6c4257387c0a712c097858)

Windows 32bit client zip: http://docker.lab99.org.s3-website-ap-southeast-2.amazonaws.com/builds/Windows/i386/docker-1.14.0-dev.zip (sha256sum: 45a1d18b1329e92064848af91af0a342a1c522eb6baf529ded1add587b896ced)

---

The line is quite long, I do prefer the previous version.

@mstanleyjones Although I prefer markdown format, I cannot use it for the link, as @tianon pointed out that the download information will be posted on multiple places, such as mailing list, where markdown format doesn't apply.

LK4D4 commented:
ping @thaJeztah @mstanleyjones

mstanleyjones commented:
I'm OK with the last iteration, but I'm only tangentially involved in this so I shouldn't really be a gatekeeper on it.

tonistiigi commented:
@vieux @thaJeztah Is this conflicting with latest version/release updates.

thaJeztah commented:
let me /cc @andrewhsu on that :blush:

twang2218 commented:
@tonistiigi It shouldn't conflict with the new version schema, as version is passed as a variable `$VERSION`, however, I rebased and tested by following command:

```bash
docker run -e AWS_S3_BUCKET=docker.lab99.org \
             -e AWS_ACCESS_KEY_ID \
             -e AWS_SECRET_ACCESS_KEY \
             -e AWS_DEFAULT_REGION \
             -e SKIP_RELEASE_BUILD=0 \
             -it --privileged \
             docker-dev:add-release-checksum \
             ./hack/release.sh --release-regardless-of-test-failure
```

and I got the following error message:

```bash
Building release

error: .git directory missing and DOCKER_GITCOMMIT not specified
  Please either build with the .git directory accessible, or specify the
  exact (--short) commit hash you are building using DOCKER_GITCOMMIT for
  future accountability in diagnosing build issues.  Thanks!
```

It seems https://github.com/docker/docker/pull/30290 removed the `.git` directory without adding `ENV`/`ARG` of `DOCKER_GITCOMMIT` to Dockerfile, which might caused the problem. @vdemeester

twang2218 commented:
After manually given the `DOCKER_GITCOMMIT` environment variable,

```bash
docker run -e AWS_S3_BUCKET=docker.lab99.org \
             -e AWS_ACCESS_KEY_ID \
             -e AWS_SECRET_ACCESS_KEY \
             -e AWS_DEFAULT_REGION \
             -e SKIP_RELEASE_BUILD=0 \
             -e DOCKER_GITCOMMIT=a996066 \
             -it --privileged \
             docker-dev:add-release-checksum \
             ./hack/release.sh --release-regardless-of-test-failure
```

the generated results are:

---

Release complete; see http://docker.lab99.org.s3-website-ap-southeast-2.amazonaws.com
Use the following text to announce the release:

We have just pushed 17.04.0-dev to http://docker.lab99.org.s3-website-ap-southeast-2.amazonaws.com. You can download it with the following:

Linux 64bit tgz: http://docker.lab99.org.s3-website-ap-southeast-2.amazonaws.com/builds/Linux/x86_64/docker-17.04.0-dev.tgz (sha256sum: 07f2200f5bb4f57d1fa763e89a0820b48de119548ceed7861819721a4cf71fd9)

Linux 32bits arm tgz: http://docker.lab99.org.s3-website-ap-southeast-2.amazonaws.com/builds/Linux/armel/docker-17.04.0-dev.tgz (sha256sum: d725534593aa282802ebb1760595ec5871f03e5c65c4d7a557d0e0970b5c31af)

Darwin/OSX 64bit client tgz: http://docker.lab99.org.s3-website-ap-southeast-2.amazonaws.com/builds/Darwin/x86_64/docker-17.04.0-dev.tgz (sha256sum: a7e65bbe807d8a9d4f4bc688fcd9d7dc82847e66ce039993690eb82336499132)

Windows 64bit zip: http://docker.lab99.org.s3-website-ap-southeast-2.amazonaws.com/builds/Windows/x86_64/docker-17.04.0-dev.zip (sha256sum: d9e6e42222fdd39b0bb61ce87f67686f7ee71227102110dc79174d9019165254)

Windows 32bit client zip: http://docker.lab99.org.s3-website-ap-southeast-2.amazonaws.com/builds/Windows/i386/docker-17.04.0-dev.zip (sha256sum: 94a9ef78c87719d59b1b26ed03254d5c80d9bfbfcc49c4b57f40b57d2233c600)

tonistiigi commented:
ping @vieux @andrewhsu 

The output LGTM

tonistiigi commented:
ping @vieux @andrewhsu

cpuguy83 commented:
moby/moby won't have releases anymore (these are done via the docker-ce repo). I think we can close this?

AkihiroSuda commented:
I think we can close this, but please consider carrying this to Docker product side?
@andrewhsu
